### ISSUE TITLE
Fix CS0133 and change from "const" to "var"

### ISSUE THREAD

**Issue by giggio** [code-cracker/code-cracker#685]
`CS0133` will be raised if you are assigning to a constant. This fix will change `const <type>` to `var`.

This:

````csharp
const string myStr = "abc".ToUpper();
````

Becomes:

````csharp
var myStr = "abc".ToUpper();
````

Category: `Compiler` (same as `CC0133`, has to be added to `SupportedCategories.cs`)

**Comment by giggio** [comment]
@code-cracker/owners can we set this to ready?

**Comment by jwooley** [comment]
If it is a class level constant, it might make more sense to use

`readonly string myStr = "abc".ToUpper();`

I'm assuming here you are trying to offer a code fix to the existing CS0133 or creating our own analyzer for the cases where const assigns to something other than a const.

**Comment by carloscds** [comment]
@jwooley I agree with you, in class level this fix will create an error. Change to readonly, in this case, is more approriate.

**Comment by giggio** [comment]
@jwooley The idea is to fix the existing `CS0133`, not create a new analyzer and diagnostic id.

@carloscds I have updated the description to enable 2 different scenarios: field and local variable.

**Comment by carloscds** [comment]
@giggio Ok

**Comment by kidchenko** [comment]
Hi, good night.

Are you working on it? If no, I want to help and maybe this issue can be a smart start for me. Thanks :+1:

**Comment by giggio** [comment]
@kidchenko go for it.

**Comment by Poojith** [comment]
Hi @giggio, I'm a first-timer with OSS. Is the issue resolved? If no, I would love to take it up.

**Comment by carloscds** [comment]
@Poojith Go ahread!

**Comment by Poojith** [comment]
Sorry for my ignorance, this is my first time with Visual Studio 2015 -- had previously worked with 2010.
While setting up the solution after adding the startup projects, I always get an error that an 'output type of class library' cannot be started. Should I be adding references to the DLLs? 
Also, please let me know if there is a mailing/contact list where I can ask my queries. Thanks!

**Comment by carloscds** [comment]
@Poojith Do you have Visual Studio 2015 with Update 3 ?

**Comment by Poojith** [comment]
@carloscds Yes, I do. I have all the required setup in place, it's just that I am facing a problem while running the project. Please let me know if there is another forum/email where I can continue this conversation. Don't want to digress away from the topic of the issue :)

**Comment by carloscds** [comment]
@Poojith Ok, open another Issue and explain error in detail, if possible with screen shot.

**Comment by carloscds** [comment]
@Poojith Wait! reading again your first post, the problem is with Startup Project. Set VSIX Project as startup project and not CodeCracker library.

**Comment by Poojith** [comment]
@carloscds thanks a lot, that helped.
I just went through the entire codebase over the past 2 hours and noticed that none of the constant variables are not changed dynamically. Pretty much all the value assignments are done in test classes. The other constant variables which are in actual code(not test cases) are not changed with methods. Please tell me if the statement from the description of the issue :

> Category: Compiler (same as CC0133, has to be added to SupportedCategories.cs)

meant the static variables had to be changed. I guess that's not the intention, though.

**Comment by carloscds** [comment]
@Poojith Can you send a sample about this ?

**Comment by Poojith** [comment]
@carloscds , apologize for the late reply.
For instance, in StringBuilderInLoopAnalyzer.cs, the fields `Title`, `MessageFormat`, or `Category `are all assignments involving string literals, and are not changed dynamically.  
I'm wondering if the description of the issue, 

> Category: Compiler (same as CC0133, has to be added to SupportedCategories.cs)

meant the static variables in `SupportedCategories.cs` had to be changed.

**Comment by sasanknvs** [comment]
Hi @carloscds , I am a first-timer in OSS. Is this issue still open ? If so, can i try to fix it ?

**Comment by carloscds** [comment]
Hi @sasanknvs, you can. Go ahead!

**Comment by sasanknvs** [comment]
@carloscds  Can you tell what is the .sln file i need to use ? I could multiple .sln files.

**Comment by sasanknvs** [comment]
@carloscds I was able to build the solution properly after installing microsoft.net.compilers. But i am not getting the issue mentioned. 
Can anyone please throw some light on it ?

**Comment by carloscds** [comment]
@sasanknvs Can you try using NuGet Packages ?

**Comment by MrZeka** [comment]
Hello @carloscds is this still open? Is anyone working on this? Is the ticket change or any progress happened in this one? @Poojith?

**Comment by carloscds** [comment]
@RobertPecz Yes, you can go ahead.

**Comment by RobertPecz** [comment]
@carloscds Okay for example the CatchEmptyAnalyzer.cs has some const fields like _'internal const string Title'_. I can't change this to readonly like this, because it giving back an compile error (CS0236 - A field initializer cannot reference the nonstatic field, method, or property). Is it okay to put this in a private constructor?

**Comment by carloscds** [comment]
@RobertPecz Can you explain more ? Are you using CatchEmptyAnalyzer as a sample ?

**Comment by RobertPecz** [comment]
@carloscds I've just checked couple of .cs files and find couple of internal const string. Yes, for example in the CatchEmptyAnalyzer.cs you can find couple of internal const string. Because of this variables are fields I should change it to readonly, but if I change this const files to readonly, I facing with a compile error - _CS0236 - A field initializer cannot reference the nonstatic field, method, or property_.
So in this situation, I can change const fields to readonly, but I have to assings its value in the constructor.

**Comment by carloscds** [comment]
@RobertPecz These fields are readonly by definition in project.

**Comment by Fabioh** [comment]
Hi @carloscds.
this is my first time in OSS and I have never worked in a compiler.
I heard  about this project in the Lambda 3 podcast  [#05 - Open Source](https://soundcloud.com/lambdatres/5-open-source).

I have cloned the project and I want to start to solve this issue.
Where can I start?

**Comment by MaStr11** [comment]
@Fabioh 

> Where can I start?

I assume you have read the [start-working](https://github.com/code-cracker/code-cracker#start-working) and pull-request sections on the  main page. 

Make sure you have the Roslyn SDK installed. [Instructions for VS 2015 and 2017](https://github.com/dotnet/roslyn-sdk)

After that you should familiarize yourself with creating analyzers. You should read this [post ](https://msdn.microsoft.com/en-us/magazine/dn879356.aspx).

What you want is a CodeFixer. You need do derive [from the `CodeFixProvider` class](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/CodeFixProvider.cs,0ea452288308834d) and add an `[ExportCodeFixProvider]` attribute. An example of such a registration can be found [here](https://github.com/dotnet/roslyn/blob/bb0aefb2aaed62c2ff5498a30ba4cfe7d9034a1b/src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs). You should get to the point were you can debug your analyzer by placing a breakpoint in `RegisterCodeFixesAsync`. Write a unit-test to get there. The fix it self requires some syntax transformations. You can read about *Working with syntax* [here](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/). There are several ways to do a syntax transformation (Note: Don't take the approach in CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider). In your case the easiest one (ReplaceNode) will likely be sufficient. You will also need some semantic analysis (GetSemanticModel) to get the type if you need to change `var` to `string`.

Once you got it working you may also consider contributing this fix to Roslyn itself. There is an issue that is tracking fixes like this one in [roslyn](https://github.com/dotnet/roslyn/issues/23326). This way you could reach a wider audience because every IDE driven by Roslyn would get you fixer.

**Comment by Fabioh** [comment]
@MaStr11 thank to clarifying  you my doubts.
I will try to follow these steps and start to fix this issue.

**Comment by MaStr11** [comment]
@Fabioh You should also have a look at #835. The *Files changed* sections contains all the files you need to change/create also (Note: It contains a *DiagnosticAnalyzer* which **creates** a diagnostic. That is something you don't need because you rely on a compiler diagnostic and you don't have to create one on your own).

**Comment by Fabioh** [comment]
Thanks a lot.
